source/Ship.cpp

```diff
@@ -3750,6 +3750,25 @@ void Ship::DoGeneration()
 	// maximum capacity for the rest of the turn, but must be clamped to the
 	// maximum here before they gain more. This is so that, for example, a ship
 	// with no batteries but a good generator can still move.
+
+	// If some or all excess energy is due to "fuel energy", then run
+	// "fuel consumption" in reverse to "retcon" the energy generation.
+
+	// Of future note: The assumption that fuel was, in fact, burned last frame,
+	// is not always true. If a ship runs out of fuel, but still has an energy
+	// surplus, it will actually backfill its fuel with excess energy, until it
+	// has enough to burn again.
```

Comment:
Why not keep track of how much fuel was generated by fuel consumption? If we're aware of an issue with this approach, I'd say just resolve it now.

Down by where fuel consumption is done:
```c++
		if(attributes.Get("fuel consumption") <= fuel)
		{
			fuel -= attributes.Get("fuel consumption");
			gainedFuelEnergy = attributes.Get("fuel energy");
			energy += gainedFuelEnergy;
			heat += attributes.Get("fuel heat");
		}
		else
			fuelEnergy = 0.;
```
Then here we'd check `if(!isDisabled && gainedFuelEnergy)`.